### ISSUE TITLE
fix(checker): Tcl/iRules script bodies are reported as string literals

### ISSUE THREAD

**Issue by dekobon** [dekobon/big-code-analysis#1381]
## Summary

`Checker::is_string` calls a Tcl / iRules `braced_word` a string literal
whatever role it plays, so `bca find --type string` reports a `proc`
body, an `if` body and an iRules `when` handler as string literals.

#1318 built the predicate that settles this — a braced word is a value
when the command it is passed to takes a value, and a script otherwise
— but it could not apply it here, because `Checker::is_string` takes
neither `code` nor `ancestors`.

## Evidence

At the pinned grammars, after #1318:

```text
$ cat p.tcl
proc p {x} { puts $x }

$ bca find --type string --no-config -p p.tcl
== p.tcl ==
╰─ {braced_word:88} from (1, 12) to (1, 23) : { puts $x }
```

That is a `proc` body — the block whose commands the walk descends into
— reported as a string literal. The literal half is already right: after
#1318 the same command reports `{a b}` and `{c d}` for
`lappend x {a b}` / `puts {c d}`, which are strings. Only the script
role is wrong.

`src/checker/tcl.rs` and `src/checker/irules.rs` both carry

```rust
impl_simple_is_string!(Tcl, QuotedWord, BracedWord, BracedWordSimple);
```

and `BracedWord` is the script kind. `Alterator::alterate` lists the
same three.

## Why this is not a Tcl-only edit

```rust
fn is_string(node: &Node) -> bool
```

is the trait signature (`src/checker.rs`), and `impl_simple_is_string!`
generates it for most of the twenty-odd languages. Deciding a Tcl braced
word's role needs the source bytes (the enclosing command's leading
word) and the ancestor chain, so honouring it means widening the
signature — either by adding an `is_string_with_code` sibling with a
forwarding default, the shape `Getter::get_op_type_with_code` and
`Getter::get_space_kind_with_code` already use, or by widening
`is_string` itself and updating every impl and call site.

Note the `_with_code` trap in `.claude/rules/grammar-dispatch.md` §7:
if a sibling method is added, every call site on the walk must use the
`_with_code` spelling, or the override silently never fires.

## Scope

- Decide between a `_with_code` sibling and a signature widening, and
  enumerate the call sites either way.
- Route Tcl and iRules `BracedWord` through the #1318 role predicate
  (`Getter::is_value_braced_word`), which currently lives on `Getter`
  and would need to be reachable from `Checker` or hoisted.
- Cross-walk `Alterator::alterate`, which lists the same kinds
  (grammar-dispatch §7).
- Pin the answer with a `bca find --type string` test per dialect: a
  script body must not be reported, and the `lappend x {a b}` literal
  that is reported today must stay reported.

## Related

- #1318 — the Halstead half, fixed; its `get_op_type` arm in
  `src/getter/tcl.rs` carries a pointer to this issue.
- #1354, #1317, #1314 — the earlier braced-word fixes.

---

## Resolution

**Status:** Fixed (pending merge) — `41809053` on
`worktree-agent-a8dd410aebf08a0d8`. (The hash may be rewritten when the
branch is re-signed at merge.)

**Root cause:** `braced_word` serves as both a script block and a
quoted literal, and `Checker::is_string` is a kind table that cannot
separate the two. `Alterator::alterate` had the same arm list and the
same defect, flattening a script body into a single leaf and dropping
the whole body from `Ast::dump` and the REST `/ast` endpoint.

**Shape:** additive on the classification side — a
`Checker::is_string_with_code` sibling with a forwarding default, plus
an `Alterator::keeps_children` veto, both resolving the role through
one shared `Getter::is_braced_script_word` over one
`lang_helpers::{tcl,irules}::BRACED_WORD_KINDS` table per dialect.

**Correction to the Scope section above:** a `_with_code` sibling alone
*is* enough to change what the CLI sees. `Filter<'a>` already carries a
lifetime so its predicates can capture the parser's source buffer — the
pre-existing `"function"` arm does exactly that — so the bytes reach a
predicate through the closure rather than through `Filter::any`'s
signature.

**What did force a signature change was cost, not correctness.** Both
role questions are about the enclosing command, so both need an
ancestor chain; off `Ancestors::unknown()` each lookup is
`Node::parent`'s `O(depth)`, which made `count --type string` and
`Ast::dump` quadratic in nesting depth (7.4 ms vs 434 ms, and 10 ms vs
418 ms, on 8 KB of nested braces). `find`, `count` and the dump walk now
thread the chain, so `Filter::any` and `Alterator::get_ast_node` each
take one more argument. Neither is covered by the stability contract —
`STABILITY.md` places the whole `big-code-analysis-ast` crate outside
it — and the root crate re-exports neither.

**Alterator cross-walk: fixed, not deferred.** Leaving it would have
put `alterate` in direct contradiction with `is_string_with_code` on
the same bytes. A braced value still flattens; a braced script keeps its
children.

**Filed while fixing this, both pre-existing:**

- #1416 — `bca find` is quadratic on *nested matches* because it
  re-renders their source text. Independent of any predicate:
  `find --type braced_word` costs the same.
- #1415 — every row of `alterator_string_flattening.rs` panics under a
  partial feature set (8 of 11 rows predate this fix).

#1382, the operand half, is untouched and stays open.

**Comment by dekobon** [comment]
Fixed on `worktree-agent-a8dd410aebf08a0d8` in `41809053` (16 files,
+612/-89). Not closing — the branch is unmerged.

## Premise confirmed

Reproduced independently at the pinned grammars, both dialects, before
touching anything:

```text
$ bca find --type string --no-config -p p.tcl        # proc p {x} { puts $x }
╰─ {braced_word:88} from (1, 12) to (1, 23) : { puts $x }

$ bca find --type string --no-config -p p.irule      # when HTTP_REQUEST { log local0. "hi" }
╰─ {braced_word:132} from (1, 19) to (3, 2)
╰─ {quoted_word:150} from (2, 17) to (2, 21) : "hi"
```

After:

```text
$ bca find --type string --no-config -p p.tcl
(nothing)

$ bca find --type string --no-config -p p.irule
╰─ {quoted_word:150} from (2, 17) to (2, 21) : "hi"

$ bca find --type string --no-config -p v.tcl        # lappend x {a b} / puts {c d}
╰─ {braced_word:88} from (1, 11) to (1, 16) : {a b}
╰─ {braced_word:88} from (2, 6) to (2, 11) : {c d}

$ bca find --type string --no-config -p v.irule      # lappend b {x y}
╰─ {braced_word:132} from (1, 11) to (1, 16) : {x y}
```

The literal half is untouched; only the script role moved. `bca ops` and
`bca metrics` are byte-identical — this changes no metric.

## The seam

The issue's scope note says a `_with_code` sibling alone "changes
nothing the CLI sees", because `Filter::any` takes only a node. That is
not quite right: **`Filter<'a>` already carries a lifetime precisely so
its predicates can borrow the parser's source buffer**, and the
pre-existing `"function"` arm already closes over `code`. So the bytes
reach a predicate through the closure, not through the signature, and
the correctness half needed no signature change at all.

What did need one was the *cost* — see below.

| File | Change |
| --- | --- |
| `checker.rs` | new `Checker::is_string_with_code`, forwarding default |
| `parser.rs` | the `"string"` filter arm calls it, closing over `code` |
| `getter.rs` | new shared `Getter::is_braced_script_word` = the script kind *and* not `is_value_braced_word` |
| `lang_helpers/{tcl,irules}.rs` | the two `BRACED_WORD_KINDS` tables hoisted here from the `Getter` impls, `pub(crate)` |
| `checker/{tcl,irules}.rs` | the two overrides |
| `alterator.rs` | new `Alterator::keeps_children` veto + the two overrides |
| `ast.rs`, `find.rs`, `count.rs` | the three walks thread an ancestor chain |

Call sites enumerated workspace-wide before changing anything: there is
exactly **one** production caller of `Checker::is_string` (the
`parser.rs` filter arm) across the root, `-cli`, `-web`, `-py` and
`fuzz/`; everything else is the 24 `impl_simple_is_string!` /
`impl_js_family_is_string!` definitions and tests. `fuzz/` only passes
the filter *name*.

## The part the issue does not mention: this was a quadratic

Both questions — "is this braced word a value" and "must this node keep
its children" — are about the *enclosing command*, so both need an
ancestor chain. Off `Ancestors::unknown()` each lookup is
`Node::parent`'s `O(depth)`, and `braced_word` is every Tcl block and
list literal, so the candidate set is dense. That is the #1052 / #1122
shape, and `node.rs` warns about it in as many words.

Measured by swapping only the `Ancestors` argument, same release build
either side:

| depth | `count --type string` threaded | unknown | `Ast::dump` threaded | unknown |
|---|---|---|---|---|
| 500 | 3.4 ms | 18.2 ms | 0.5 ms | 7.0 ms |
| 1000 | 3.7 ms | 34.6 ms | 1.7 ms | 26.9 ms |
| 2000 | 4.4 ms | 118.2 ms | 3.5 ms | 105.4 ms |
| 4000 | **7.4 ms** | **433.9 ms** | **10.4 ms** | **417.8 ms** |

Linear against quadratic; 58.6x and 40.2x at depth 4000 on an 8 KB
input. The fix is the one `parser.rs` had already written down since
#1162 — "giving `find`/`count` the `(node, depth)` stack would supply a
known chain" — so `find`, `count` and the dump walk now maintain one,
with the same truncate/push discipline `spaces::compute::metrics_inner`
uses and `Ancestors::checked` guarding it.

`Filter::any` and `Alterator::get_ast_node` take the chain as a new
argument. **APIBREAK: none against the published contract** —
`STABILITY.md` places the whole `big-code-analysis-ast` crate, naming
`Filter` and `Ancestors` explicitly, outside the shape contract, and
the root crate re-exports neither.

Separately, `bca find` (not `count`) is *still* quadratic on nested
matches, because it re-renders the nested source text of each one. That
is a pre-existing defect with nothing to do with this predicate —
`find --type braced_word`, a single `kind_id` comparison, costs the
same 408 ms — and is filed as #1416.

## The `Alterator` decision: fixed

`Alterator::alterate` listed `BracedWord` beside `QuotedWord` and
`BracedWordSimple` and flattened all three into a childless leaf. For a
script body that discarded the entire body from `Ast::dump` and the REST
`/ast` endpoint: every command inside a `proc` vanished behind one leaf
holding its text, and an iRules file collapsed to one leaf per handler.

Leaving it would have put the alterator in direct contradiction with
`is_string_with_code` on the same node, which is what §7 forbids, and
the arm's own comment says it exists to agree with `is_string`. It now
takes the same predicate through an `Alterator::keeps_children` veto: a
braced **value** still flattens (Tcl evaluates nothing between its
braces, so the value is its whole span), a braced **script** keeps its
children.

The veto is a separate hook rather than an `alterate` arm so that
`alterate` keeps its byte-and-kind signature — only this question needs
the chain, and threading it through all twenty-odd `alterate` impls to
serve two of them would make every language pay for one grammar's
ambiguity.

A side effect worth naming: this makes the iRules `quoted_word`
flattening arm reachable at all. `irules_flattens_braced_word` asserted
that a `when` body came back as one leaf, so the only verbatim text it
could look for was the brace block; in valid iRules a quoted word only
ever appears inside a handler body, so that arm had no reachable input.
The row is now `irules_flattens_quoted_word`.

## Tests

- `tcl_script_body_is_a_string_only_to_the_byteless_spelling`
  (`checker.rs`) — the direct predicate test, asserting the two
  spellings **disagree** on a script body and **agree** on every other
  kind, and that they answer identically off a known chain and off a
  `Node::parent` climb. It reports the disputed nodes by *text*, not by
  count: a count is polarity-blind, and an inverted predicate satisfied
  the count form.
- `find_string_reports_tcl_family_literals_and_not_script_bodies` and
  `count_string_agrees_with_find_on_tcl_family_bodies`
  (`tests/api/ast_seam_test.rs`) — per dialect, asserting the exact
  reported list rather than the body's absence.
- Four rows in `alterator_string_flattening.rs`: a braced value still
  flattens in each dialect, and a quoted word inside a script body is
  reachable in each.

All are `#[cfg(any(feature = "tcl", feature = "irules"))]` on the `fn`
with a `ran > 0` guard; verified absent — not failing — under
`--features go`, and each dialect alone runs its own row.

Verified by a thirteen-perturbation sweep covering the production lines
*and* the fixtures. Every perturbation fails at least one test except
the chain threading, which is behaviour-preserving by construction (an
unknown chain gives the same answers and only costs more); that one is
covered by the measurement above plus `Ancestors::checked`, whose
`debug_assert!` I confirmed live by desynchronising each of the three
walks in turn.

Three gaps the sweep found and closed, none of which a passing suite
showed: deleting the **Tcl** alterator arm failed nothing (the existing
Tcl row reached its literal at statement level, where the veto has
nothing to do); inverting the predicate's polarity failed nothing in the
checker test; and the Tcl seam fixture could lose its `proc` entirely
with no test noticing.

## Not in scope

- #1382 (the operand half) — untouched, still open.
- #1416 — `bca find`'s rendering quadratic, pre-existing.
- #1415 — `alterator_string_flattening.rs` rows panic under a partial
  feature set, pre-existing (8 of 11 rows predate this fix).

## Gate

`BCA_GATE: pass (gate=pre-commit)`; `make chain-audit` exits 0
(3337 + 337 tests) with the exact `chain.last() == node.parent()`
assertion over all three new chain walks; all nine CI `feature-matrix`
legs pass.
